Store VM anonymous classes to shared cache

sogutbera opened the pull request:
1. Changed the check for lambda class names and hash keys to check only before the index number
2. Added a romSize check at first to avoid checking everything else
3. Changed the check for class file size because size might differ when index numbers have different numbers of digits
4. Changed shouldCheckForEquality to not check the class names for lambda classes
5. Added tests
6. Added a compareROMClassForEquality Trace Event for tests
7. Added the command line option -XX:[+/-]ShareAnonymousClasses
8. Disabled shared anonymous classes for a test to not store extra orphans
9. Added a test for disabling shared lambdas
10. Added the description of the command line option to the output of the help option

Closes: #5665

Signed-off-by: sogutbera <sogutbera@ibm.com>

hangshao0 commented:
Please fix the copyright failures. You need to update the end date in the copyright of each modified file to 2019. 

Also you need only one signed-off in the commit message. Example of commit message can be found here:
https://github.com/eclipse/openj9/blob/master/CONTRIBUTING.md

hangshao0 commented on runtime/bcutil/ComparingCursor.cpp:
```diff
@@ -109,7 +110,17 @@ ComparingCursor::writeU32(U_32 u32Value, DataType dataType)
+		 * difference is less than sizeof(U_32) to match lambda classes that have
+		 * different sized class names (e.g., Test1$$Lambda$9/0000000000000000 and Test1$$Lambda$10/0000000000000000)
+		 */
+		if(CLASS_FILE_SIZE == dataType && isLambda()){
```
Coding standard is here https://github.com/eclipse/omr/blob/master/doc/CodingStandard.md. There should be a space between "if" and "(", ")" and "{". Please fix this in all the lines you changed.

hangshao0 commented on runtime/bcutil/ComparingCursor.cpp:
```diff
@@ -416,12 +428,20 @@ ComparingCursor::shouldCheckForEquality(DataType dataType, U_32 u32Value)
+	if(isLambda()){
+		/* if the class is a lambda class we compare everything except the class name */
+		if(SRP_TO_UTF8_CLASS_NAME == dataType)
+			return false;
```
Can you try move this to Line 444. We skip SRP_TO_UTF8_CLASS_NAME if isLambda() is true. For all other types, we return the same as the non-lambda class.

hangshao0 commented on runtime/bcutil/ROMClassBuilder.cpp:
```diff
@@ -266,6 +266,11 @@ ROMClassBuilder::handleAnonClassName(J9CfrClassFile *classfile)
 	PORT_ACCESS_FROM_PORT(_portLibrary);
 
+    /* check if the class is a lambda class */
+    if(NULL != getEndOfLambdaClassName(originalStringBytes, originalStringLength)){
```
Incorrect indent ?

hangshao0 commented on runtime/bcutil/ROMClassBuilder.cpp:
```diff
@@ -424,13 +429,14 @@ ROMClassBuilder::prepareAndLaydown( BufferManager *bufferManager, ClassFileParse
 	 */
 	Trc_BCU_Assert_False(context->isRetransforming() && !context->isRetransformAllowed());
 
+    bool isLambda = false;
```
Incorrect indent ?

hangshao0 commented on runtime/bcutil/ROMClassBuilder.cpp:
```diff
@@ -1281,15 +1295,27 @@ bool
+		/* if it is a lambda class we already compared the romSizes so we assign the romSize of the compared class to match them */
+		romSize = ((J9ROMClass *)romClass)->romSize;
+	}
+	else Trc_BCU_Assert_True(romSize == 0);
```
"}" and "else" should be on the same line.

hangshao0 commented on runtime/bcutil/ROMClassBuilder.cpp:
```diff
@@ -1281,15 +1295,27 @@ bool
+
+	if(isLambda){
+		/* if it is a lambda class we already compared the romSizes so we assign the romSize of the compared class to match them */
+		romSize = ((J9ROMClass *)romClass)->romSize;
```
If shouldCheckForEquality() returns false for ROM_SIZE, there is no need to set romSize here.

hangshao0 commented on runtime/jcl/common/sun_misc_Unsafe.cpp:
```diff
@@ -100,6 +100,11 @@ Java_sun_misc_Unsafe_defineAnonymousClass(JNIEnv *env, jobject receiver, jclass
 	jsize length = env->GetArrayLength(bytecodes);
 
+	if(NULL != vm->sharedClassConfig){
+		/* if shared classes are enabled for an anonymous class, the class loader shared classes should be enabled to store the class in the cache */
```
Better to mention isROMClassShareable() here in this comment that we are storing lamda classes to the shared cache.

hangshao0 commented on runtime/oti/util_api.h:
```diff
@@ -2197,6 +2197,11 @@ getJVMFeature(J9JavaVM *vm);
+/**
+ * @return pointer to the last '$' sign of the class name if it is a lambda class, NULL otherwise
+ */
+char* getEndOfLambdaClassName(const char *className, UDATA classNameLength);
```
Probably a better name for this function is getLastDollarSignOfLambdaClassName().

hangshao0 commented on runtime/util/shchelp_j9.c:
```diff
@@ -86,3 +86,31 @@ getOpenJ9Sha()
 }
+
+/**
+ * @return pointer to the last '$' sign of the class name if it is a lambda class, NULL otherwise
```
Please add doxygen comment for each input parameter.

hangshao0 commented on runtime/util/shchelp_j9.c:
```diff
@@ -86,3 +86,31 @@ getOpenJ9Sha()
+ */
+char*
+getEndOfLambdaClassName(const char *className, UDATA classNameLength){
+
```
Better to check if className is NULL or classNameLength is 0

hangshao0 commented on test/functional/cmdLineTests/shareClassTests/SCCMLTests/ShareClassesCMLTests-1.xml:
```diff
@@ -36,11 +36,13 @@
 	<variable name="CACHE_DIR" value="cacheDir=."/>
 	
 	<variable name="currentMode" value="$mode204$"/>
+	<variable name="lambdaMode" value="-Xtrace:print={j9shr.2259}"/>
```
No need to add lambdaMode here. You can directly add -Xtrace:print={j9shr.2259} to the commands

hangshao0 commented on test/functional/cmdLineTests/shareClassTests/SCCMLTests/ShareClassesCMLTests-1.xml:
```diff
@@ -1049,6 +1051,120 @@
+	
+	<test id="Test 59: Make sure that when the program runs again lambda classes are used from the cache and not stored again " timeout="600" runPath=".">
+	    <command>$JAVA_EXE$ $currentMode$ $lambdaMode$ $CP_HANOI$ $PROGRAM_LAMBDA$ 0</command>
+	    <output type="success" caseSensitive="yes" regex="no">Lambda test done!</output>
```
Probably we can add a trace point that indicates the return value inside ROMClassBuilder::compareROMClassForEquality() and check that trace point here.

hangshao0 commented on test/functional/cmdLineTests/utils/src/org/openj9/test/lambdatests/Test1.java:
```diff
@@ -0,0 +1,151 @@
+package org.openj9.test.lambdatests;
+
+import java.util.Comparator;
+
```
Please add comment here explaining what is being tested for each case.

hangshao0 commented:
> Changed the check for 4 orphans in the test to 4 or more orphans because there could be more than 4 now

This test case in shrcdbgextddrtests.xml has recent captured a bug for me. I would like to keep this test case (starting up org.openj9.test.ivj.Hanoi twice with -Xshareclass resulting in 4 orphans). We can add this case in any of the ShareClassesCMLTests (Using -Xshareclasses:noBootclasspath). This can be done via a separate PR.

hangshao0 commented on runtime/bcutil/ComparingCursor.cpp:
```diff
@@ -52,7 +52,8 @@ ComparingCursor::ComparingCursor(J9JavaVM *javaVM, SRPOffsetTable *srpOffsetTabl
 	_varInfoHelper(srpOffsetTable, romClass, context),
-	_isEqual(true)
+	_isEqual(true),
+	_isLambda(isLambda)
```
I think a better name for this filed can be something like isCompareLamdaFromSCC. We might have code outside shared classes using ComparingCursor in the future.

sogutbera commented:
> > Changed the check for 4 orphans in the test to 4 or more orphans because there could be more than 4 now
> 
> This test case in shrcdbgextddrtests.xml has recent captured a bug for me. I would like to keep this test case (starting up org.openj9.test.ivj.Hanoi twice with -Xshareclass resulting in 4 orphans). We can add this case in any of the ShareClassesCMLTests (Using -Xshareclasses:noBootclasspath). This can be done via a separate PR.

This is not possible because even when using noBootclasspath I am seeing these lambda classes being stored in the cache and we do not have an option to disable storing lambda classes

`12:00:55   [ERR] 1: 0x00007F3596807F98 ORPHAN: jdk/internal/module/SystemModuleFinders$SystemModuleReader$$Lambda$1/0000000000000000 at 0x00007F3585459000.
12:00:55   [ERR] 1: 0x00007F3596807F28 ORPHAN: jdk/internal/loader/BuiltinClassLoader$$Lambda$2/0000000000000000 at 0x00007F3585459418.
12:00:55   [ERR] 1: 0x00007F3596807F18 ORPHAN: org/openj9/test/ivj/Hanoi at 0x00007F3585459808.
12:00:55   [ERR] 1: 0x00007F3596807DF8 ORPHAN: org/openj9/test/ivj/Post at 0x00007F358545A2D8.
12:00:55   [ERR] 1: 0x00007F3596807DC4 ORPHAN: org/openj9/test/ivj/Disk at 0x00007F358545A688.`

hangshao0 commented:
> This is not possible because even when using noBootclasspath I am seeing these lambda classes being stored in the cache and we do not have an option to disable storing lambda classes

OK. I see.

pshipton commented on runtime/util/shchelp_j9.c:
```diff
@@ -86,3 +86,41 @@ getOpenJ9Sha()
+	/* Get the pointer to the last '$' sign */
+	char* end = strnrchrHelper(className, '$', classNameLength);
+
+	if (NULL != end && (end - 8) - className >= 0) {
```
Please bracket the sub-expressions here, in the if statement above, and the one below.
`if ((NULL != end) && ((end - 8) - className >= 0))`

pshipton commented:
> and we do not have an option to disable storing lambda classes

I think it's a good idea to add an option to control the behavior. Although rather than an option to control storing the lambda classes, I'd like an option to avoid reading them from the shared cache. Or adding options to control both behavior's is fine. These can be used for diagnostics. 

`-XX:[+/-]ReadSharedLambdas`
`-XX:[+/-]WriteSharedLambdas`

Please also do some performance testing on this change so we can see how it affects startup time.

pshipton commented on runtime/util/shchelp_j9.c:
```diff
@@ -86,3 +86,41 @@ getOpenJ9Sha()
+		   'd' == *(end - 2) &&
+		   'a' == *(end - 1)
+		) {
+			/* Check if $$Lambda$ exists in the class name */
```
Technically it's only checking for `$$Lambda` and not `$$Lambda$`. Either the code of the comment needs to be changed.

hangshao0 commented:
If we are adding new command line options, we need new documentation for these options. To request documentation change, an issue can be created at: https://github.com/eclipse/openj9-docs/issues. FYI @sogutbera

sogutbera commented:
> > and we do not have an option to disable storing lambda classes
> 
> I think it's a good idea to add an option to control the behavior. Although rather than an option to control storing the lambda classes, I'd like an option to avoid reading them from the shared cache. Or adding options to control both behavior's is fine. These can be used for diagnostics.
> 
> `-XX:[+/-]ReadSharedLambdas`
> `-XX:[+/-]WriteSharedLambdas`
> 
> Please also do some performance testing on this change so we can see how it affects startup time.

https://github.com/eclipse/openj9/blob/927509892bffd47e2496fcd7a9f6313f67dd2a9a/runtime/bcutil/ROMClassBuilder.cpp#L513
If we disable isROMClassShareable, both read and write will get disabled, so if we want a behaviour that only disables write, we could add an if that checks if WriteSharedLambdas is enabled after this line
https://github.com/eclipse/openj9/blob/927509892bffd47e2496fcd7a9f6313f67dd2a9a/runtime/bcutil/ROMClassBuilder.cpp#L573 that covers until this line https://github.com/eclipse/openj9/blob/927509892bffd47e2496fcd7a9f6313f67dd2a9a/runtime/bcutil/ROMClassBuilder.cpp#L629

if we want a behaviour that only disables read, we could skip storing the lambda class in the hash table or return false for compareROMClassForEquality or also add an if that checks if ReadSharedLambdas is enabled before comparing.

pshipton commented:
A single option that disabled both read and write should be fine. How about `-XX:[+/-]ShareLambdas`.

sogutbera commented:
> Please also do some performance testing on this change so we can see how it affects startup time.

I did a performance test and there is no impact on the performance.

sogutbera commented:
> A single option that disables both read and write should be fine. How about `-XX:[+/-]ShareLambdas`.

I have fixed the things that you mentioned in your comments and I also added this option and it seems to work. Could you please review again?

hangshao0 commented:
You may want to add an test case for -XX:-ShareLambdas in ShareClassesCMLTests-1.xml

sogutbera commented:
> You may want to add an test case for -XX:-ShareLambdas in ShareClassesCMLTests-1.xml

I added it.

DanHeidinga commented:
Is the `-XX:-ShareLambdas` name accurate?  Does this only store Lambda-related classes or any class allocated by Unsafe.defineAnonymousClass?

sogutbera commented:
> Is the `-XX:-ShareLambdas` name accurate? Does this only store Lambda-related classes or any class allocated by Unsafe.defineAnonymousClass?

I changed the name to -XX:[+/-]ShareAnonymousClasses. Can you please review my changes?

hangshao0 commented on runtime/shared/shrclssup.c:
```diff
@@ -137,6 +137,13 @@ IDATA J9VMDllMain(J9JavaVM* vm, IDATA stage, void* reserved)
+				/* Check for -XX:+ShareAnonymousClasses and -XX:-ShareAnonymousClasses; whichever comes later wins. Enable is set by default so we just need to disable when that's the case. */
+				argIndex1 = FIND_ARG_IN_VMARGS(OPTIONAL_LIST_MATCH, VMOPT_XXENABLESHAREANONYMOUSCLASSES, NULL);
+				argIndex2 = FIND_ARG_IN_VMARGS(OPTIONAL_LIST_MATCH, VMOPT_XXDISABLESHAREANONYMOUSCLASSES, NULL);
+				if(argIndex2 > argIndex1){
```
Missing space between if and "(". ) and "{"

hangshao0 commented on runtime/nls/exel/exelib.nls:
```diff
@@ -889,3 +889,17 @@ J9NLS_EXELIB_INTERNAL_HELP_XXSHARED_CACHE_HARD_LIMIT_EQUALS.explanation=NOTAG
 J9NLS_EXELIB_INTERNAL_HELP_XXSHARED_CACHE_HARD_LIMIT_EQUALS.user_response=
 # END NON-TRANSLATABLE
+
+J9NLS_EXELIB_INTERNAL_HELP_XXENABLESHAREANONYMOUSCLASSES=\  -XX:+ShareAnonymousClasses   enable storing anonymous classes in the shared class cache
```
should be enable storing and finding

hangshao0 commented on runtime/nls/exel/exelib.nls:
```diff
@@ -889,3 +889,17 @@ J9NLS_EXELIB_INTERNAL_HELP_XXSHARED_CACHE_HARD_LIMIT_EQUALS.explanation=NOTAG
+# END NON-TRANSLATABLE
+
+J9NLS_EXELIB_INTERNAL_HELP_XXDISABLESHAREANONYMOUSCLASSES=\  -XX:-ShareAnonymousClasses   disable storing anonymous classes in the shared class cache
+# START NON-TRANSLATABLE
```
should be "disable storing and finding"

hangshao0 commented on runtime/jcl/common/sun_misc_Unsafe.cpp:
```diff
@@ -100,6 +100,13 @@ Java_sun_misc_Unsafe_defineAnonymousClass(JNIEnv *env, jobject receiver, jclass
 	jsize length = env->GetArrayLength(bytecodes);
 
+	if (NULL != vm->sharedClassConfig) {
+		if(vm->sharedClassConfig->runtimeFlags & J9SHR_RUNTIMEFLAG_ENABLE_SHAREANONYMOUSCLASSES){
```
Please use `if (J9_ARE_ALL_BITS_SET(vm->sharedClassConfig->runtimeFlag, J9SHR_RUNTIMEFLAG_ENABLE_SHAREANONYMOUSCLASSES)) {`

sogutbera commented:
Fixed everything. @DanHeidinga can you please take a look? Thanks.

DanHeidinga commented on runtime/util/shchelp_j9.c:
```diff
@@ -86,3 +86,41 @@ getOpenJ9Sha()
+	/* Get the pointer to the last '$' sign */
+	char* end = strnrchrHelper(className, '$', classNameLength);
+
+	if ((NULL != end) && ((end - 8) - className >= 0)) {
```
I think this needs to `>` not `>=` as there won't be a valid lambda with the name `$$Lambda$`  It needs to have at least a single character before the $$ for the host

DanHeidinga commented on runtime/util/shchelp_j9.c:
```diff
@@ -86,3 +86,41 @@ getOpenJ9Sha()
+	char* end = strnrchrHelper(className, '$', classNameLength);
+
+	if ((NULL != end) && ((end - 8) - className >= 0)) {
+		if (('$' == *(end - 8)) &&
```
A `memcmp(end-8, "$$Lambda", sizeof("$$Lambda") -1)` might be clearer.

sogutbera commented on runtime/bcutil/ROMClassBuilder.cpp:
```diff
@@ -1281,18 +1295,48 @@ bool
+	bool ret = false;
+
+	if (isLambda) {
+		if (sizeof(U_64) < abs((int)(romSize - ((J9ROMClass *)romClass)->romSize))) {
```
The HostClassName$$Lambda$ will be the same for both classes at both of these checks because hash key is checked before. So the only difference for the sizes will be the difference between the number of digits of the index number. 

romSize increases by 2 when the number of digits of the index number increases by 1, but romSize has padding and is always multiples of 8. So this romSize check will allow at least 4 but up to 7 difference in the number of digits of the index number. 

classFileSize increases by 1 when the number of digits of the index number increases by 1 and it has no padding so this check will allow exactly 4 digits difference. 

So we would be allowing 4 digits difference in any case which means about 100000 lambda classes under the same host class. (eg. HostClassName$$Lambda$[1-9] can get matched up to HostClassName$$Lambda$99999, HostClassName$$Lambda$[10-99] can get matched up to HostClassName$$Lambda$999999)

DanHeidinga replied:
Can you add some of this additional explanation to the code so we don't lose the detials on why they are different checks?

sogutbera replied:
I added some additional explanations to both of the checks. Could you take a look? Thanks.

hangshao0 commented:
I guess we need to add doc label to this PR (or #5665).

DanHeidinga commented:
> I guess we need to add doc label to this PR (or #5665).

Done

DanHeidinga commented:
FYI @dsouzai - will additional AOT work be required for this to get the startup benefits?  In particular, the SCC will be doing a bit of a fuzzy match on the anon class names to find the right one which I'm unclear on how that will work with class chains

dsouzai commented:
I don't think there needs to be any additional AOT work. The class chains work entirely on ROM Classes, eg:

https://github.com/eclipse/openj9/blob/5dd23af8483d4671fe6ad2ff372bcf0e6fd74891/runtime/compiler/env/J9SharedCache.cpp#L604-L616

When we do validations, we always go from the RAM Class to the ROM Class; we get the RAM Class from some (runtime) VM query, and then check its class chain, which first involves getting its ROM Class and then finding the Class Chain from the ROM Class.

@mstoodle am I missing any subtlety?

DanHeidinga commented:
FYI @mpirvu & @vijaysun-omr for awareness of the AOT discussions above

mstoodle commented:
I suggest looking carefully at the AOT class loader heuristic to make sure it works well with multiple anonymous classes. It is largely based on class chains, so I believe it could work but I'd like someone to work through it to make sure it all hangs together.

@dsouzai It would be nice to see some tests to verify that a shared lambda can be successfully inlined into an AOT compiled method and then loaded and validated in a second run :) .

DanHeidinga commented:
@sogutbera & @hangshao0 - have any startup performance tests been done on this change?  I'd like to see Liberty Startup benchmarks run before this is merged.

sogutbera commented:
> > Please also do some performance testing on this change so we can see how it affects startup time.
> 
> I did a performance test and there is no impact on the performance.

@DanHeidinga I did a performance test on both CleanedLibertyStartupDT(17dev-4way-0-256-qs) and CleanedLibertyStartupDT7(17dev-4way-0-256-qs) and there was no impact on the performance.

DanHeidinga commented:
> It would be nice to see some tests to verify that a shared lambda can be successfully inlined into an AOT compiled method and then loaded and validated in a second run :) .

@sogutbera Have you been able to validate this already with your existing tests?  Seeing a lambda get loaded from the SCC and that the AOT was found would be a big plus.

sogutbera commented:
> > It would be nice to see some tests to verify that a shared lambda can be successfully inlined into an AOT compiled method and then loaded and validated in a second run :) .
> 
> @sogutbera Have you been able to validate this already with your existing tests? Seeing a lambda get loaded from the SCC and that the AOT was found would be a big plus.

I just tried with my test and the AOT was found for some of the lambdas. One of the outputs: 
`+ (AOT load) jdk/internal/module/SystemModuleFinders$SystemModuleReader$$Lambda$1/0000000000000000.get$Lambda(Ljdk/internal/module/SystemModuleFinders$SystemModuleReader;)Ljava/util/function/Function; @ 00000000223572E0-0000000022357475 Q_SZ=0 Q_SZI=0 QW=1 j9m=000000000E488788 bcsz=9 compThread=0Found AOT code for ROMMethod 0x0000000032325708 in shared cache.`

I also tried running Eclipse and I'm also seeing a lot of AOT loads for lambdas. Here is one of the outputs: 
`+ (AOT load) java/util/regex/Pattern$$Lambda$70/0000000000000000.<init>()V @ 000000002CD7C640-000000002CD7C72F Q_SZ=0 Q_SZI=0 QW=1 j9m=0000000018AB1498 bcsz=5 compThread=0Found AOT code for ROMMethod 0x000000003CACED30 in shared cache.`

vijaysun-omr commented:
Could be quite timely. I see the following from a javacore running Quarkus with OpenJ9 (the classes do not end up in the SCC yet but obviously we expect that to change with this change):

``` 2CLTEXTCLLOAD  		Loader java/lang/InternalAnonymousClassLoader(0x00000000F80A7FA8)
3CLTEXTCLASS   			org/jboss/resteasy/core/ServerResponseWriter$$Lambda$69/000000009800F0A0(0x00000000010DDC00)
3CLTEXTCLASS   			org/jboss/resteasy/core/SynchronousDispatcher$$Lambda$68/000000009800EB20(0x00000000010DD700)
3CLTEXTCLASS   			org/jboss/resteasy/core/ResourceMethodInvoker$$Lambda$67/000000009800A580(0x00000000010DD200)
3CLTEXTCLASS   			org/jboss/resteasy/core/ResourceMethodInvoker$$Lambda$66/000000009800A280(0x00000000010D6700)
3CLTEXTCLASS   			org/jboss/resteasy/core/ResourceMethodInvoker$$Lambda$65/0000000098009DE0(0x00000000010DCD00)
3CLTEXTCLASS   			org/jboss/resteasy/core/MethodInjectorImpl$$Lambda$64/0000000098008F80(0x00000000010DC800)
3CLTEXTCLASS   			org/jboss/resteasy/core/MethodInjectorImpl$$Lambda$63/0000000098008BD0(0x00000000010D6200)
3CLTEXTCLASS   			org/jboss/resteasy/core/MethodInjectorImpl$$Lambda$62/00000000980087F0(0x00000000010D5E00)
3CLTEXTCLASS   			org/jboss/resteasy/core/MethodInjectorImpl$$Lambda$61/00000000980081A0(0x00000000010D5800)
3CLTEXTCLASS   			org/jboss/resteasy/core/ResourceMethodInvoker$$Lambda$60/0000000098007BB0(0x00000000010D5400)
3CLTEXTCLASS   			org/jboss/resteasy/core/ResourceMethodInvoker$$Lambda$59/0000000098006DC0(0x00000000010D4F00)
3CLTEXTCLASS   			org/jboss/resteasy/plugins/server/resourcefactory/POJOResourceFactory$$Lambda$58/00000000980069D0(0x00000000010D4A00)
3CLTEXTCLASS   			org/jboss/resteasy/plugins/server/resourcefactory/POJOResourceFactory$$Lambda$57/00000000980063C0(0x00000000010D4500)
3CLTEXTCLASS   			org/jboss/resteasy/core/SynchronousDispatcher$$Lambda$56/0000000098005F60(0x00000000010D4000)
3CLTEXTCLASS   			org/jboss/resteasy/core/SynchronousDispatcher$$Lambda$55/0000000098005740(0x00000000010D3B00)
3CLTEXTCLASS   			io/quarkus/arc/EventImpl$$Lambda$54/0000000010310500(0x00000000010DB700)
3CLTEXTCLASS   			sun/net/ExtendedOptionsImpl$$Lambda$53/0000000094002BB0(0x00000000010DB300)
3CLTEXTCLASS   			java/util/stream/Collectors$$Lambda$52/00000000104381A0(0x0000000001100800)
3CLTEXTCLASS   			java/util/stream/Collectors$$Lambda$51/0000000010437EE0(0x0000000001100300)
3CLTEXTCLASS   			java/util/stream/Collectors$$Lambda$50/00000000102DFBB0(0x00000000010FFE00)
3CLTEXTCLASS   			java/util/stream/Collectors$$Lambda$49/00000000102DF740(0x00000000010FFA00)
3CLTEXTCLASS   			io/quarkus/undertow/runtime/UndertowDeploymentTemplate$$Lambda$48/00000000102DF420(0x00000000010FF500)
3CLTEXTCLASS   			org/xnio/nio/NioXnioWorker$NioWorkerMetrics$$Lambda$47/00000000102DEF30(0x00000000010FF000)
3CLTEXTCLASS   			org/xnio/XnioWorker$$Lambda$46/0000000010376870(0x00000000010B3E00)
3CLTEXTCLASS   			org/xnio/XnioWorker$$Lambda$45/0000000010376290(0x00000000010B3A00)
3CLTEXTCLASS   			org/wildfly/common/context/ContextManager$$Lambda$44/0000000010375FC0(0x00000000010B3600)
3CLTEXTCLASS   			org/xnio/XnioWorker$$Lambda$43/0000000010375D30(0x00000000010AB100)
3CLTEXTCLASS   			io/quarkus/resteasy/runtime/RolesFilterRegistrar$$Lambda$42/00000000103683B0(0x000000000104A900)
3CLTEXTCLASS   			io/quarkus/resteasy/runtime/RolesFilterRegistrar$$Lambda$41/00000000103680C0(0x000000000104A400)
3CLTEXTCLASS   			java/util/stream/MatchOps$$Lambda$40/0000000010367C50(0x0000000001049F00)
3CLTEXTCLASS   			io/quarkus/resteasy/runtime/RolesFilterRegistrar$$Lambda$39/0000000010362F60(0x0000000001049A00)
3CLTEXTCLASS   			io/quarkus/resteasy/runtime/RolesFilterRegistrar$$Lambda$38/0000000010362B50(0x0000000001049500)
3CLTEXTCLASS   			io/quarkus/resteasy/runtime/RolesFilterRegistrar$$Lambda$37/0000000010362770(0x0000000001049100)
3CLTEXTCLASS   			java/util/stream/Collectors$$Lambda$36/0000000010362090(0x0000000001040C00)
3CLTEXTCLASS   			java/util/stream/Collectors$$Lambda$35/0000000010361540(0x0000000000F90E00)
3CLTEXTCLASS   			java/util/stream/Collectors$$Lambda$34/00000000103611D0(0x0000000000F90900)
3CLTEXTCLASS   			java/util/stream/Collectors$$Lambda$33/0000000010360F90(0x0000000000F90500)
3CLTEXTCLASS   			io/quarkus/resteasy/runtime/RolesFilterRegistrar$$Lambda$32/0000000010360AE0(0x0000000000F90000)
3CLTEXTCLASS   			io/quarkus/resteasy/runtime/RolesFilterRegistrar$$Lambda$31/000000001035FB90(0x0000000000F8FB00)
3CLTEXTCLASS   			org/jboss/resteasy/core/ResteasyProviderFactoryImpl$$Lambda$30/0000000010354DE0(0x0000000000F8ED00)
3CLTEXTCLASS   			org/jboss/resteasy/core/ResteasyProviderFactoryImpl$$Lambda$29/0000000010256810(0x0000000000F8E800)
3CLTEXTCLASS   			com/ibm/lang/management/internal/MemoryNotificationThread$$Lambda$28/00000000AC0048A0(0x0000000000F8E400)
3CLTEXTCLASS   			org/jboss/resteasy/core/ConstructorInjectorImpl$$Lambda$27/00000000102F9C00(0x0000000000D9C500)
3CLTEXTCLASS   			io/quarkus/arc/ComputingCache$CacheFunction$$Lambda$26/00000000102E9CC0(0x0000000000D9C000)
3CLTEXTCLASS   			io/quarkus/arc/BeanManagerImpl$$Lambda$25/00000000102E7580(0x0000000000D9BC00)
3CLTEXTCLASS   			io/quarkus/arc/EventImpl$ObserverExceptionHandler$$Lambda$24/000000001024EF00(0x0000000000D9B700)
3CLTEXTCLASS   			io/quarkus/arc/ArcContainerImpl$$Lambda$23/00000000102E5930(0x0000000000D9B200)
3CLTEXTCLASS   			io/quarkus/arc/ArcContainerImpl$$Lambda$22/00000000102E5560(0x0000000000D9AE00)
3CLTEXTCLASS   			io/quarkus/arc/ArcContainerImpl$$Lambda$21/00000000102E5170(0x0000000000D9A900)
3CLTEXTCLASS   			io/quarkus/arc/AbstractSharedContext$$Lambda$20/000000001024EBF0(0x0000000000D9A400)
3CLTEXTCLASS   			io/quarkus/arc/ArcContainerImpl$$Lambda$19/000000001024E8F0(0x0000000000D99F00)
3CLTEXTCLASS   			org/wildfly/common/expression/Expression$$Lambda$18/00000000102D4350(0x0000000000D99500)
3CLTEXTCLASS   			io/quarkus/runtime/configuration/ConfigExpander$$Lambda$17/00000000102D40C0(0x0000000000D99100)
3CLTEXTCLASS   			io/quarkus/runtime/configuration/ExpandingConfigSource$$Lambda$16/0000000010254670(0x0000000000D98C00)
3CLTEXTCLASS   			io/smallrye/config/ImplicitConverters$$Lambda$15/00000000102CD6E0(0x0000000000D98700)
3CLTEXTCLASS   			io/smallrye/config/Converters$$Lambda$14/00000000102C5580(0x0000000000D97E00)
3CLTEXTCLASS   			io/smallrye/config/Converters$$Lambda$13/00000000102C50C0(0x0000000000D97900)
3CLTEXTCLASS   			io/smallrye/config/Converters$$Lambda$12/00000000102C4C60(0x0000000000D97400)
3CLTEXTCLASS   			io/smallrye/config/Converters$$Lambda$11/00000000102C3C90(0x0000000000D96F00)
3CLTEXTCLASS   			io/smallrye/config/Converters$$Lambda$10/00000000102C3000(0x0000000000D96A00)
3CLTEXTCLASS   			io/smallrye/config/Converters$$Lambda$9/00000000102C2BD0(0x0000000000D96500)
3CLTEXTCLASS   			io/smallrye/config/Converters$$Lambda$8/00000000102C27A0(0x0000000000D96000)
3CLTEXTCLASS   			io/smallrye/config/Converters$$Lambda$7/00000000102C2370(0x0000000000D74C00)
3CLTEXTCLASS   			io/smallrye/config/Converters$$Lambda$6/00000000102C1F40(0x0000000000D74700)
3CLTEXTCLASS   			io/smallrye/config/Converters$$Lambda$5/0000000010221990(0x0000000000D74200)
3CLTEXTCLASS   			io/smallrye/config/SmallRyeConfigBuilder$$Lambda$4/00000000102BFD90(0x0000000000D73D00)
3CLTEXTCLASS   			java/util/function/Function$$Lambda$3/00000000102BF990(0x0000000000D73800)
3CLTEXTCLASS   			io/quarkus/runtime/configuration/ExpandingConfigSource$$Lambda$2/00000000102226E0(0x0000000000D73300)
3CLTEXTCLASS   			java/util/function/UnaryOperator$$Lambda$1/00000000102B1BD0(0x0000000000D72E00)```

DanHeidinga commented:
Jenkins test sanity plinux,xlinux jdk8

DanHeidinga commented:
Jenkins test sanity zlinux,aix jdk11

DanHeidinga commented:
Jenkins test extended win,osx jdk12

sogutbera commented:
The test SyntheticGCWorkload_concurrentSlackAuto_1M_J9_0 failed but it is due to this issue: https://github.com/eclipse/openj9/issues/6715
 It is fixed now by this PR: https://github.com/AdoptOpenJDK/openjdk-tests/pull/1283 but it failed because the last time I pushed was 10 days ago.

dsouzai commented:
> I suggest looking carefully at the AOT class loader heuristic to make sure it works well with multiple anonymous classes. It is largely based on class chains, so I believe it could work but I'd like someone to work through it to make sure it all hangs together.

The class chain verification looks ok. In terms of the class loader heuristic, I'm going to need some clarification on what happens for anonymous classes. 

For some reason I have a vague notion that an anonymous class is loaded by an anonymous class loader, but then later can be associated with a non-anonymous class loader. @DanHeidinga 1. is that true, and 2. if so, is that anonymous class still "visible" from the anonymous class loader?

The class loader heuristic makes a class loader environment-agnostic by mapping it to the first class it loaded - this is likely to happen run to run but isn't guaranteed (hence the term heuristic). In order to verify a profiled class, we'll try to get the class loader that loaded it originally and then use the romclass+classloader to get our hands on the profiled class (this is done because a profiled class isn't named in the bytecodes; we poof it up via profiling). 

So, if the anonymous class can have its class loader changed to a non-anonymous class loader, then we'll fail a bunch of validations because we won't find it when we try to look it up via the romclass+anonymous_classloader. However, if an anonymous class is always tied to its anonymous class loader, then everything's the same as the non-anonymous case. Either way though, there's no functional problem; just a performance penalty.

DanHeidinga commented:
The `InternalAnonymousClassLoader` "owns" the memory segments for the anonymous RAM / ROM Classes but the anon class's `J9Class->classLoader` field is the same as the `host` class.

The API for creating a anon class is `Unsafe.defineAnonymousClass(Class host, byte[] classfile, Object[] patches)`

The `InternalAnonymousClassLoader` can't be used to look up the classes as we never register them in them its the classHashTable as part of being anon means they can't be looked up by name.  It's purely a fiction to own the memory segments.  I guess you could find the class by walking the segments but it's not the usual path...

> Either way though, there's no functional problem; just a performance penalty.

Can you expand on the performance penalty?  Will it prevent the AOT for lambdas from being loaded?

dsouzai commented:
> The InternalAnonymousClassLoader "owns" the memory segments for the anonymous RAM / ROM Classes but the anon class's J9Class->classLoader field is the same as the host class.

Ok, so then it looks like there shouldn't be a CL problem; because:
```
static void jitHookClassLoad(J9HookInterface * * hookInterface, UDATA eventNum, void * eventData, void * userData)
   {
   J9VMInternalClassLoadEvent * classLoadEvent = (J9VMInternalClassLoadEvent *)eventData;
   J9VMThread * vmThread = classLoadEvent->currentThread;
   J9Class * cl = classLoadEvent->clazz;
   ...
   J9ClassLoader *classLoader = cl->classLoader;
   ...
   compInfo->getPersistentInfo()->getPersistentClassLoaderTable()->associateClassLoaderWithClass(classLoader, clazz);
   ...
   }
```

> Can you expand on the performance penalty? Will it prevent the AOT for lambdas from being loaded?

For the most part no; the heuristic only comes into play for classes that we got our hands on via profiling. Such a class might be used for in an guard for example. It could go two ways: either we wouldn't be able to validate the class that was used for the guard, or we'd pick the wrong RAMClass (but with the right classchain). The former means that we'd have to fail the load, and the latter means that we end up only allowing the execution of the inlined code for a class that might not be as frequent. Both scenarios are functionally ok but less performant.

DanHeidinga commented:
Only failure is the synthetic gc test on mac - a known issue.
